sidecar/app/query/base.py

```diff
@@ -77,23 +83,55 @@ def get_from_query_id(cls, query_id) -> Optional["Query"]:
             if query:
                 return query
             raise e
-        if query.status_file_path.exists():
-            with query.status_file_path.open("r") as f:
-                status_str = f.readline()
-                query.status = Status[status_str]
-                return query
-        return None
+        query.load_history_from_file()
+        if query.status == Status.UNKNOWN:
+            return None
+        return query
+
+    def load_history_from_file(self):
+        if self.status_file_path.exists():
+            self.logger.debug(
+                f"Loading query {self.query_id} status history "
+                f"from file {self.status_file_path}"
+            )
+            with self.status_file_path.open("r") as f:
+                for line in f:
+                    status_str, timestamp = line.split(",")
+                    self._status_history.append(
+                        StatusChangeEvent(
+                            status=Status[status_str], timestamp=float(timestamp)
+                        )
+                    )
+
+    @property
+    def _last_status_event(self):
+        if not self._status_history:
+            return StatusChangeEvent(status=Status.UNKNOWN, timestamp=time.time())
+        return self._status_history[-1]
+
+    @property
+    def status_event_json(self):
+        status_event = {
+            "status": self._last_status_event.status.name,
+            "start_time": self._last_status_event.timestamp,
+        }
+        if self.status >= Status.COMPLETE and len(self._status_history) >= 2:
+            status_event["start_time"] = self._status_history[-2].timestamp
+            status_event["end_time"] = self._last_status_event.timestamp
+        return status_event
 
     @property
     def status(self) -> Status:
-        return self._status
+        return self._last_status_event.status
 
     @status.setter
     def status(self, status: Status):
-        self._status = status
-        with self.status_file_path.open("w") as f:
-            self.logger.debug(f"setting status: {status=}")
-            f.write(str(status.name))
+        if self.status <= Status.COMPLETE:
+            now = time.time()
+            self._status_history.append(StatusChangeEvent(status=status, timestamp=now))
+            with self.status_file_path.open("a") as f:
```

Comment:
It may be better to have a class that manages the history and flushes it to the backing storage, so `Query` can take itself off this unrelated to its core functionality business

Reply:
yep, that's exactly what #71 does 🙂